### ISSUE TITLE
[feature] Add basic temporal handling support for vector layers

### ISSUE THREAD

**Issue by QGIS-feature-tracker** [qgis/QGIS-Documentation#5443]
Original commit: https://github.com/qgis/QGIS/commit/d98defec1a06e512cdf47e28cb659cace266e6be by nyalldawson

This exposes some basic temporal capabilities for vector layers:
- static time range for layer (to match raster layer possibilities), this
sets a single static time range which applies to the whole layer. ALL
features from the layer will be shown whenever the canvas time
overlaps the layer time range
- "Single field with datetime": Allows selection of a single date
or datetime field from the layer. Features will be shown whenever
this field value is within the canvas time range
- "Separate Fields for Start and End Date/Time": Allows selection
of start and end date/datetime fields from the layer. Features will
be shown whenever the time interval calculated from these fields
overlaps the canvas time range

Some known limitations/inefficiencies:
- currently only date/datetime fields can be used. This was done
to simplify the format handling and avoid the need to worry about
string fields with different datetime formats. In future we should
allow selection of string fields and allow users to enter a custom
datetime format string
- unlike the Time Manager plugin approach, the approach taken here
is to rely completely on QGIS expressions and feature requests to
do the filtering (Time Manager uses layer filter strings and attempts
to set a native SQL filter syntax so that filtering is done on the
backend). This is intentional, because it provides a unified filter
approach regardless of the provider used (i.e. we don't need to worry
about the different SQL syntaxes used natively by the different
providers). The beauty of feature request expression compilation
**should** mean that the QGIS expressions are magically turned into
native backend queries, BUUUUUUUUUUUT... because we lack QGIS expression
support for date time literals, we currently rely on the "to_datetime"
expression function and coerce everything through strings. None of
the expression compilers handle this function, so currently ALL
filtering is done on the QGIS side. We need to add functions for
optimised datetime literal creation, and then ensure that the different
compilers correctly map these literals across to the backend
filter syntax to allow all the filtering work to be done on the database
side...

So currently, performance is much worse with large layers compared
to Time Manager. But, the advantage is that we can use the native
temporal framework and have vector layers animated alongside mesh
and raster layers!

**Comment by ghtmtt** [comment]
I'm planning to start documenting this feature. I really think it is worth to have a nice separated section like `Temporal Control`:

* temporal control involves vector, raster and mesh
* there are a **lot** of different features and potentially a lot of coming enhancements
* having a *standalone* section allows us to not *over* fill already long files

opinions? ping @DelazJ @alexbruy @rduivenvoorde @havatv

**Comment by ghtmtt** [comment]
I was thinking a structure like the following:

```
├── temporal_control
│   ├── img
│   ├── index.rst
│   ├── intro.rst
│   ├── mesh.rst
│   ├── raster.rst
│   └── vector.rst
```

so we can split all the different providers and having a clean structure

**Comment by rduivenvoorde** [comment]
@ghtmtt Thanks for starting! I'm happy to read/help in this.

Not sure if there is enough information to have all those files. Why not just start with one file with ... an index, introduction, and then (if needed) some special cases for providers? IF that becomes too big we can always cut it in pieces?

**Comment by rduivenvoorde** [comment]
I'm going to work on the code here too, as I think that we mix up the different kinds of possible data-types and frame-types here in QGIS. 
I think(!) that you can have 
- data which have a point in time (like when you have Tweets or so)
- data which has a range in time (like measurements between 1st and 31st of januari)
Then then you can have a timeframe (that is the 'showing frame') ALSO in a point in time (like: only data of 1st of januari 2020) OR a timeframe (like 1st till 31st of januari).
And sets of those two, define if you should take the 'borders' of either timeframe OR data(frame) in to account.
I think this is currently to defined in the right way....
(not sure why I tell this here, it is just that I try to make things clear for myself I think :-) )

Also this is important to note: https://github.com/qgis/QGIS/issues/38647

**Comment by ghtmtt** [comment]
thanks for the feedback @rduivenvoorde!

> Not sure if there is enough information to have all those files. Why not just start with one file with ... an index, introduction, and then (if needed) some special cases for providers? IF that becomes too big we can always cut it in pieces?

yes, I'll start with `index.rst` and `intro.rst`. Maybe in the `intro.rst` we can add some sections for the different data providers and than cut it into pieces (if needed of course ;) ).

The data used by @nyalldawson in his *webinar* (earthquake vector data) are perfectly suitable as a starting point (I already got the data).

**Comment by rduivenvoorde** [comment]
Yeah, I also found https://earthquake.usgs.gov/earthquakes/feed/v1.0/geojson.php which you can use to get certain different type of earthquake datasets.

**Comment by DelazJ** [comment]
>> Not sure if there is enough information to have all those files. Why not just start with one file with ... an index, introduction, and then (if needed) some special cases for providers? IF that becomes too big we can always cut it in pieces?

>yes, I'll start with index.rst and intro.rst. Maybe in the intro.rst we can add some sections for the different data providers and than cut it into pieces (if needed of course ;) ).

I'm also not sure there is enough information to split the chapter. If a single file has the contents, we don't need an index page, do we? What about a single `temporal.rst` file (more expressive than intro) in which we put all the information, with sections by (I don't know what you have in mind) data types, project/options...? Or do you think there will be so much information that a single page would not be enough (and coherent)?

**Comment by ghtmtt** [comment]
let's proceed with a single `temporal.rst` file (actually the index has sense only if we have more files) and let's have single section with all the data providers. Then we will see how to proceed and if we should split the information.

BTW: I'm planning to add some gifs but I don't think we have already some of them and if they can cause some troubles. Dealing with time i could be useful to have some animations.

**Comment by DelazJ** [comment]
There was one in [2.18](https://docs.qgis.org/2.18/en/docs/user_manual/grass_integration/grass_integration.html#importing-data-into-a-grass-location-via-drag-and-drop) docs but removed in 0208f3e79fc (but I couldn't find why but my guess was "pdf" build, and looking into the corresponding doc, there was indeed no icon displayed).
Seems that this sphinx extension [Imgconverter](https://www.sphinx-doc.org/en/master/usage/extensions/imgconverter.html) would help here (https://github.com/sphinx-doc/sphinx/pull/4537/files). Worth a try....
